lib/db_connection.ex

```diff
@@ -793,7 +798,16 @@ defmodule DBConnection do
   `transaction/3` call inside another `transaction/3` will be treated
   the same as `run/3`.
 
-  ### Options
+  > #### Checkout failures {: .warning}
+  >
+  > If we cannot check out a connection from the pool, this function raises a
+  > `DBConnection.ConnectionError` exception. This function also raises a
+  > `DBConnection.ConnectionError` exception if a `DBConnection` callback
+  > invoked because of code in `fun` returns a bad value.
+  > If you want to handle these cases, you should rescue
+  > `DBConnection.ConnectionError` exceptions when using `run/3`.
```

Comment:
Two additional comments:

1. The invalid return type means an adapter is buggy, so it is always a bug upstream which should be addressed (but in theory should never happen), so I am not sure if we should mention it

2. "If you want to handle these cases" - I assume the only case of interest is "if we cannot check out a connection from the pool", right?

So my suggestion is:

```suggestion
  > `DBConnection.ConnectionError` exception.
  > If you want to handle these cases, you should rescue
  > `DBConnection.ConnectionError` exceptions when using `run/3`.
```

In other words, sure, other exceptions can happen, but they should not happen unless a bug outside of our code (and outside of the usual logic of not being able to checkout).